db/src/debug.rs

```diff
@@ -252,3 +308,168 @@ pub(crate) fn dump_sql_query(conn: &rusqlite::Connection, sql: &str, params: &[&
+        let materialized_attribute_map = read_attribute_map(&self.sqlite).expect("schema map");
+
+        let materialized_schema = Schema::from_ident_map_and_attribute_map(materialized_ident_map, materialized_attribute_map).expect("schema");
+        assert_eq!(materialized_schema, self.schema);
+    }
+
+    pub fn transact<I>(&mut self, transaction: I) -> Result<TxReport> where I: Borrow<str> {
+        // Failure to parse the transaction is a coding error, so we unwrap.
+        let entities = edn::parse::entities(transaction.borrow()).expect(format!("to be able to parse {} into entities", transaction.borrow()).as_str());
+
+        let details = {
+            // The block scopes the borrow of self.sqlite.
+            // We're about to write, so go straight ahead and get an IMMEDIATE transaction.
+            let tx = self.sqlite.transaction_with_behavior(TransactionBehavior::Immediate)?;
+            // Applying the transaction can fail, so we don't unwrap.
+            let details = transact(&tx, self.partition_map.clone(), &self.schema, &self.schema, NullWatcher(), entities)?;
+            tx.commit()?;
+            details
+        };
+
+        let (report, next_partition_map, next_schema, _watcher) = details;
+        self.partition_map = next_partition_map;
+        if let Some(next_schema) = next_schema {
+            self.schema = next_schema;
+        }
+
+        // Verify that we've updated the materialized views during transacting.
+        self.assert_materialized_views();
+
+        Ok(report)
+    }
+
+    pub fn transact_simple_terms<I>(&mut self, terms: I, tempid_set: InternSet<TempId>) -> Result<TxReport> where I: IntoIterator<Item=TermWithTempIds> {
+        let details = {
+            // The block scopes the borrow of self.sqlite.
+            // We're about to write, so go straight ahead and get an IMMEDIATE transaction.
+            let tx = self.sqlite.transaction_with_behavior(TransactionBehavior::Immediate)?;
+            // Applying the transaction can fail, so we don't unwrap.
+            let details = transact_terms(&tx, self.partition_map.clone(), &self.schema, &self.schema, NullWatcher(), terms, tempid_set)?;
+            tx.commit()?;
+            details
+        };
+
+        let (report, next_partition_map, next_schema, _watcher) = details;
+        self.partition_map = next_partition_map;
+        if let Some(next_schema) = next_schema {
+            self.schema = next_schema;
+        }
+
+        // Verify that we've updated the materialized views during transacting.
+        self.assert_materialized_views();
+
+        Ok(report)
+    }
+
+    pub fn last_tx_id(&self) -> Entid {
+        self.partition_map.get(&":db.part/tx".to_string()).unwrap().index - 1
+    }
+
+    pub fn last_transaction(&self) -> edn::Value {
+        transactions_after(&self.sqlite, &self.schema, self.last_tx_id() - 1).expect("last_transaction").0[0].into_edn()
+    }
+
+    pub fn datoms(&self) -> edn::Value {
```

Comment:
Can we make sure that at the end of these patches we also have `pub fn transactions(&self) ...`?  It's somewhere in these stacks of patches.